standards/core/rln-contract.md

```diff
@@ -0,0 +1,301 @@
+| Maximum total rate limit of all memberships | `R_{max}` | `20000` | messages per `epoch`               |
+| Minimal rate limit of one membership        | `r_{min}` | `20`    | messages per `epoch`               |
+| Price of `1` message per epoch              | `p_u`     | `0.01`  | `USD` per one period of length `T` |
+| Membership expiration term                  | `T`       | `90`    | days                               |
+| Membership grace period                     | `G`       | `30`    | days                               |
+
+## Membership lifecycle
+
+Any existing membership MUST always be in exactly one of the following states:
+- _Active_;
+- _GracePeriod_;
+- _Expired_;
+- _ErasedAwaitsWithdrawal_;
+- _Erased_.
+
+```mermaid
+graph TD;
+    NonExistent --> |"register"| Active;
+    Active --> |"time T passed"| GracePeriod;
+    GracePeriod --> |"extend"| Active;
+    GracePeriod --> |"time G passed"| Expired;
+    GracePeriod --> |"withdraw"| Erased;
+    Expired --> |"withdraw"| Erased;
+    Expired --> |"another membership reuses slot"| ErasedAwaitsWithdrawal;
+    ErasedAwaitsWithdrawal --> |"withdraw"| Erased;
+
+```
+
+State updates triggered by a transaction (e.g., from _GracePeriod_ to _Active_ as a result of `extend`) MUST be applied immediately.
+State updates defined by time progression (e.g., from _GracePeriod_ to _Expired_ after time `G`) MAY be applied lazily.
+
+When providing any membership-specific functionality, the contract MUST:
+- check whether the state of the membership involved is up-to-date;
+- if necessary, update the membership state;
+- process the transaction in accordance with the up-to-date membership state.
+
+Memberships MUST be included in the RLN tree according to the following table:
+
+| State                    | Included in the RLN tree |
+| ------------------------ | ------------------------ |
+| _Active_                 | Yes                      |
+| _GracePeriod_            | Yes                      |
+| _Expired_                | Yes                      |
+| _ErasedAwaitsWithdrawal_ | No                       |
+| _Erased_                 | No                       |
+
+Memberships MUST NOT be transferable.
+A user MAY use one Ethereum address to manage multiple memberships.
+A user MAY use one Waku node to manage multiple memberships. [^1]
+
+[^1]: No Waku implementation supports managing multiple memberships from one node (as of August 2024).
+
+
+## Contract functionalities
+
+Availability of membership-specific functionalities MUST be as follows:
+
+|                       | Active | GracePeriod | Expired | ErasedAwaitsWithdrawal | Erased |
+| --------------------- | ------ | ----------- | ------- | ---------------------- | ------ |
+| Send a message        | Yes    | Yes         | Yes     | No                     | No     |
+| Extend the membership | No     | Yes         | No      | No                     | No     |
```

Comment:
I wondered if we should allow extending when expired.
I think not, because it introduces possible transaction failure when 2 transactions come in at the same time, one to extend and one to register on the expired slot.
ie, current proposal sounds good.

Reply:
I agree with your conclusion but I'm not sure that

> it introduces possible transaction failure when 2 transactions come in at the same time

is a relevant argument here.

Every state with multiple outgoing edges can incur a race condition when transitions to different states are triggered at around the same time. For example: a user tries to `withdraw` one second before the grace period `G` expires. Or: a user tries to `withdraw` from an `Expired` membership, and a new registrations comes in that tries to overtake that slot.

I don't see the race condition itself as an issue here. After all, the underlying blockchain is there exactly to resolve race conditions.